File: src/main/java/no/novari/flyt/history/model/time/ManualTimeFilter.java
package no.novari.flyt.history.model.time;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import no.novari.flyt.history.validation.MinTimestampBeforeMaxTimestamp;
import org.springframework.format.annotation.DateTimeFormat;

import java.time.OffsetDateTime;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@MinTimestampBeforeMaxTimestamp
public class ManualTimeFilter {

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    private OffsetDateTime min;

    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    private OffsetDateTime max;

}

File: src/main/kotlin/no/novari/flyt/history/model/time/ManualTimeFilter.kt
package no.novari.flyt.history.model.time

import no.novari.flyt.history.validation.MinTimestampBeforeMaxTimestamp
import org.springframework.format.annotation.DateTimeFormat
import java.time.OffsetDateTime

@MinTimestampBeforeMaxTimestamp
data class ManualTimeFilter(
    @field:DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    val min: OffsetDateTime? = null,
    @field:DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
    val max: OffsetDateTime? = null,
) {
    companion object {
        @JvmStatic
        fun builder() = Builder()
    }

    class Builder {
        private var min: OffsetDateTime? = null
        private var max: OffsetDateTime? = null

        fun min(min: OffsetDateTime?) = apply { this.min = min }

        fun max(max: OffsetDateTime?) = apply { this.max = max }

        fun build() = ManualTimeFilter(min = min, max = max)
    }
}
